src/Configuration.ts

```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright 2020-2023 Delft University of Technology and SynTest contributors
+ *
+ * This file is part of SynTest Framework - SynTest Core.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Yargs = require("yargs");
+import * as path from "path";
+import shell = require("shelljs");
+
+export let CONFIG: ArgumentsObject;
+export class Configuration {
+  initializeConfigSingleton<T extends ArgumentsObject>(argumentValues: T) {
+    if (CONFIG) {
+      throw Error("Already initialized the config singleton!");
+    }
+
+    CONFIG = argumentValues;
+  }
+
+  private programName: string;
+
+  constructor(programName: string) {
+    this.programName = programName;
+  }
+
+  loadConfigurationFile(cwd?: string) {
+    cwd = cwd || process.env.SYNTEST_CWD || process.cwd();
+    const configPath = path.join(cwd, ".syntest.js");
+    let config;
+    // Catch syntestjs syntax errors
+    if (shell.test("-e", configPath)) {
+      try {
+        config = require(configPath);
+      } catch (error) {
+        throw new Error(error);
+      }
+      // Config is optional
+    } else {
+      config = {};
+    }
+
+    return config;
+  }
+
+  processArguments<O extends OptionsObject>(
+    yargs: O,
+    args: string[]
+  ): ArgumentsObject {
+    const config = this.loadConfigurationFile();
+    const configuredArgs = yargs.config(config);
+    return configuredArgs.wrap(configuredArgs.terminalWidth()).parseSync(args);
+  }
+
+  configureOptions() {
+    const yargs = Yargs.usage(`Usage: ${this.programName} [options]`)
+      .example(
+        `${this.programName} -r ./src`,
+        "Running the tool with target directory 'src'"
+      )
+      .example(
+        `${this.programName} -r ./src --population_size 10`,
+        "Setting the population size"
+      )
+      .epilog("visit https://syntest.org for more documentation");
+
+    const generalOptions = this.configureGeneralOptions(yargs);
```

Comment:
It might make more sense to call all these variables `options`. It is not clear why we keep passing the different configs into the following functions. We can still derive what config it is from the function name.

This would also make the last line `researchModeOptions.showHidden(false)` less confusing. Currently, it looks like we are hiding the research options.

Reply:
I agree, that readability wise that would be better. I did try to do it that way at first but then the type inference does not work sadly. if we use a single options object like this:
let options = configureA(...)
options = configureB(options)
...

Then the final options object would have the type based on configureA only.

I added comments to clarify a bit.

Reply:
Okay, then in that case, can we just number them as is done in other places in the code (e.g., options1, option2, etc)